# Table of Contents

1. [Select Stocks](#select-stocks)
2. [Scrapling](#scrapling)
3. [Yahoo API]()
4. [Selenium](#selenium)
5. [Playwright](#playwright)
6. [Perplexity Results](#perplexity-results-gpt-4o)
   - 3.1 [Prompt](#prompt)
   - 3.2 [Response](#response)
   - 3.3 [Cointegration](#cointegration)
   - 3.4 [Correlation](#correlation)
7. [Cointegration](#cointegration)
8. [Sentiment](#sentiment)
9. [Trade Results](#trade-results)
10. [Overall Summary](#overall-summary)
11. [Breakdown Summary](#breakdown-summary)
12. [Future Works](#future-works)


# POC

Methodology for POC:

1. Choose "AAPL".
2. Scrape news (from Yahoo Finance), for Apple stock via scrapling
3. Use ChatGPT to determine the sentiment of each news on scale of 1 to 5:

- 1 -> Negative
- 2 -> Moderate negative
- 3 -> Neutral
- 4 -> Moderate positive
- 5 -> Positive

4. Use ChatGPT to determine the stock that has highest cointegration with Apple

5. Use Selenium to load more news article


In [2]:
import pandas as pd
import random
import sys
from pathlib import Path
import scrapling
from scrapling import Fetcher
from selenium import webdriver
from selenium.webdriver.common.by import By
import time
from bs4 import BeautifulSoup
import re
from datetime import datetime, timedelta
import yfinance as yf
import numpy as np
from collections import Counter
import math
from pprint import pformat
from decimal import Decimal
from scipy import stats

repo_dir = Path.cwd().as_posix()

if repo_dir not in sys.path:
    sys.path.append(repo_dir)

%load_ext autoreload
%autoreload 2

from src.utils import utils, yahoo_utils
from src.gen_data import GenData
from src.sentiment import SentimentRater
# from src.cointegrate import CoIntegrate
from src.gen_price_action import GenPriceAction
from src.cal_profit_loss import StockTrade
from src.cal_profit_loss import CalProfitLoss
from src.analyze_coint import AnalyzeCoint


The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


# [Select Stocks](#table-of-contents)

- Test "AAPL", "NVDA", and "PG" for POC.
- If time permits, then test the randomly selected stocks for each GICS Sector: ['PWR', 'INCY', 'CDW', 'CEG', 'BAC', 'LIN', 'BWA', 'ARE', 'CHTR', 'KDP', 'PSX']


In [33]:
stock_list = utils.gen_stock_list()
stock_list

{'Industrials': 'PWR',
 'Health Care': 'INCY',
 'Information Technology': 'CDW',
 'Utilities': 'CEG',
 'Financials': 'BAC',
 'Materials': 'LIN',
 'Consumer Discretionary': 'BWA',
 'Real Estate': 'ARE',
 'Communication Services': 'CHTR',
 'Consumer Staples': 'KDP',
 'Energy': 'PSX'}

In [36]:
utils.get_gics_sector(
    ["AAPL", "MSFT", "NVDA", "GOOGL", "META", "XOM", "CRM", "HD", "PG"]
)

{'AAPL': 'Information Technology',
 'MSFT': 'Information Technology',
 'NVDA': 'Information Technology',
 'GOOGL': 'Communication Services',
 'META': 'Communication Services',
 'XOM': 'Energy',
 'CRM': 'Information Technology',
 'HD': 'Consumer Discretionary',
 'PG': 'Consumer Staples'}

# [Scrapling](#table-of-contents)

- https://finance.yahoo.com/quote/AAPL/news/
- Scrapling aren't able to web scrap fully since the news are loaded when user scroll to the end of screen.


In [98]:
# Extract news related to AAPL from Yahoo Finance
news_info = yahoo_utils.extract_news_briefs(
    "https://finance.yahoo.com/quote/AAPL/news/"
)
news_info

[2025-03-08 21:01:29] INFO: Fetched (200) <GET https://finance.yahoo.com/quote/AAPL/news/> (referer: https://www.google.com/search?q=yahoo)


[{'publisher': 'Motley Fool',
  'period': datetime.timedelta(seconds=7200),
  'title': 'Could Apple Be a Millionaire-Maker Stock?',
  'content': 'During its rise to become a dominant tech enterprise, Apple (NASDAQ: AAPL) has certainly made its early investors some serious money. In the past two decades, shares have generated a total return of greater than 18,000%, boosted by the introduction of popular hardware devices and services. Could Apple be a millionaire-maker stock?'},
 {'publisher': 'Insider Monkey',
  'period': datetime.timedelta(seconds=43200),
  'title': 'Apple Inc. (AAPL) Unveils AI-Powered MacBook Air with M4 Chip',
  'content': 'We recently compiled a list of the 10 Hot AI News Updates Analysts are Monitoring. In this article, we are going to take a look at where Apple Inc. (NASDAQ:AAPL) stands against the other AI stocks. According to Goldman Sachs, the AI-powered rally in Chinese equities could boost other markets as well. Consequently, the firm has […]'},
 {'publisher

# [Yahoo API](#table-of-contents)

- Yahoo API gives very detailed news as info but limited to max 10 news article.


In [ ]:
ticker = yf.Ticker("AAPL")
ticker.news

[{'id': 'ee08c516-d812-48ab-9787-b73c34782bf7',
  'content': {'id': 'ee08c516-d812-48ab-9787-b73c34782bf7',
   'contentType': 'STORY',
   'title': "Wall Street's souring view on the US stock market isn't just about the economy — it's about AI too",
   'description': '',
   'summary': "As economic uncertainty has spawned a stock market sell-off, one of investors' favorite trades of the past two years is going through its own rerating of expectations.",
   'pubDate': '2025-03-11T17:44:49Z',
   'displayTime': '2025-03-11T17:44:49Z',
   'isHosted': True,
   'bypassModal': False,
   'previewUrl': None,
   'thumbnail': {'originalUrl': 'https://s.yimg.com/os/creatr-uploaded-images/2025-03/88bc1fd0-fb15-11ef-8dbd-ea4aea580f37',
    'originalWidth': 4500,
    'originalHeight': 3001,
    'caption': '',
    'resolutions': [{'url': 'https://s.yimg.com/uu/api/res/1.2/CAOKVOjTmaybJSdSciq_vQ--~B/aD0zMDAxO3c9NDUwMDthcHBpZD15dGFjaHlvbg--/https://s.yimg.com/os/creatr-uploaded-images/2025-03/88bc1fd0-fb1

# [Selenium](#table-of-contents)

- Selenium is not suitable as it occasionally run into ReadTimeoutError causing the code to break. Hence we are not able to perform scrolling to required level.
- Running in Jupyter Notebook causes timeout error due to async operation of Jupyter Notebook.


In [ ]:
url = "https://finance.yahoo.com/quote/AAPL/news/"
scroll_limit = 5

driver = webdriver.Chrome()
driver.get(url)

# Locate <a> tag and navigate to its parent <div>
parent_div = driver.find_elements(By.XPATH, "//div[@class='content yf-82qtw3']")

# parent_div.get_attribute("class")

# # Print the class attribute of the parent <div>
# title_element = parent_div.find_element(By.XPATH, ".//h3[contains(@class, 'yf-82qtw3)]")
# title_element.text

ReadTimeoutError: HTTPConnectionPool(host='localhost', port=37797): Read timed out. (read timeout=120)

# [Playwright](#table-of-contents)

- Able to perform scrolling to extract required number of articles as html content.
- Use BeautifulSoup on the generated html content to extract the relevant HTML tags.


In [ ]:
<div class="content yf-82qtw3"><a class="subtle-link fin-size-small titles noUnderline yf-1xqzjha"
        data-ylk="elm:hdln;elmt:link;itc:0;ct:story;slk:Shareholders%20are%20showing%20signs%20of%20DEI%20fatigue%20as%20activists%20push%20for%20more%20votes;sec:qsp-news;subsec:all;cpos:1;g:a42c9cba-3739-4dae-8402-e5acaa0ad944"
        href="https://finance.yahoo.com/news/shareholders-are-showing-signs-of-dei-fatigue-as-activists-push-for-more-votes-140136546.html"
        aria-label="Shareholders are showing signs of DEI fatigue as activists push for more votes"
        title="Shareholders are showing signs of DEI fatigue as activists push for more votes" data-rapid_p="23"
        data-v9y="1">
        <h3 class="clamp  yf-82qtw3">Shareholders are showing signs of DEI fatigue as activists push for more votes</h3>
        <p class="clamp  yf-82qtw3">Champions and critics of DEI policies are again pushing American companies to either
            bolster or diminish their diversity policies. But so far this year, none of the proposals are garnering
            shareholder support.</p>
    </a>
    <div class="footer yf-82qtw3">
        <div class="publishing yf-1weyqlp">Yahoo Finance <i aria-hidden="true" data-svelte-h="svelte-sg6bkb">•</i> 13
            hours ago</div>
        <div class="taxonomy-links yf-13tszyc"><a data-testid="ticker-container"
                class="ticker x-small hover2 border streaming yf-1fqyif7" aria-label="DE"
                data-ylk="elm:qte;elmt:link;itc:0;sec:qsp-news;slk:DE;subsec:all;cpos:1" href="/quote/DE/" title="DE"
                data-rapid_p="24" data-v9y="1">
                <div class="name yf-1fqyif7"><span class="symbol yf-1fqyif7">DE </span> </div> <fin-streamer
                    class="percentChange yf-1fqyif7" data-symbol="DE" data-field="regularMarketChangePercent"
                    data-trend="txt" data-pricehint="2" data-tstyle="default" active=""><span
                        class="d60f3b00 c4af00a5">+3.48%</span></fin-streamer>
            </a> <a data-testid="ticker-container" class="ticker x-small hover2 border streaming yf-1fqyif7"
                aria-label="COST" data-ylk="elm:qte;elmt:link;itc:0;sec:qsp-news;slk:COST;subsec:all;cpos:1"
                href="/quote/COST/" title="COST" data-rapid_p="25" data-v9y="1">
                <div class="name yf-1fqyif7"><span class="symbol yf-1fqyif7">COST </span> </div> <fin-streamer
                    class="percentChange yf-1fqyif7" data-symbol="COST" data-field="regularMarketChangePercent"
                    data-trend="txt" data-pricehint="2" data-tstyle="default" active=""><span
                        class="d60f3b00 d5882186">-6.07%</span></fin-streamer>
            </a> <a data-testid="ticker-container" class="ticker x-small hover2 border streaming yf-1fqyif7"
                aria-label="AAPL" data-ylk="elm:qte;elmt:link;itc:0;sec:qsp-news;slk:AAPL;subsec:all;cpos:1"
                href="/quote/AAPL/" title="AAPL" data-rapid_p="26" data-v9y="1">
                <div class="name yf-1fqyif7"><span class="symbol yf-1fqyif7">AAPL </span> </div> <fin-streamer
                    class="percentChange yf-1fqyif7" data-symbol="AAPL" data-field="regularMarketChangePercent"
                    data-trend="txt" data-pricehint="2" data-tstyle="default" active=""><span
                        class="d60f3b00 c4af00a5">+1.59%</span></fin-streamer>
            </a> </div>
    </div>
</div>

# [Perplexity Results (GPT-4o)](#table-of-contents)


## [Prompt](#table-of-contents)

The list of dictionaries are attached with the following prompt in Perplexity:

```
You are a stock analysis expert.

I've attached list of dictionaries containing news info specifically:

1. publisher = Source where news are generated
2. period = Amount of time lapsed after news release
3. title = News title
4. content = News content in brief

Each dictionary is supposed to be related to Apple stock.

Your task:
1. Review the 'title' and 'content' keys in each dictionary to access the sentiment related to Apple stock.
2. Use ONLY following rating:

    1 = Negative
    2 = Moderate negative
    3 = Neutral
    4 = Moderate positive
    5 = Positive

3. If the news are not related to a Apple stock, please give a rating of 3.
4. Please review all dictionary items in the attached list and provide the corresponding rating.
5. After reviewing all dictionary items, kindly provide an overall sentiment rating for Apple stock.
```


## [Response](#table-of-contents)

Response from Perplexity (ChatGPT 4-o):

```
Sentiment Ratings for Each News Item
Title: "Could Apple Be a Millionaire-Maker Stock?"
Content: Highlights Apple's historical success and potential for future gains.
Rating: 5 (Positive)

Title: "Apple Inc. (AAPL) Unveils AI-Powered MacBook Air with M4 Chip"
Content: Discusses the launch of an innovative product, indicating progress in AI.
Rating: 5 (Positive)

Title: "Apple (AAPL) Beats Stock Market Upswing: What Investors Need to Know"
Content: Reports a positive trading session for Apple stock.
Rating: 4 (Moderate positive)

Title: "Apple Delays Siri Upgrade Indefinitely as AI Concerns Escalate"
Content: Focuses on delays in Apple's AI projects, raising concerns.
Rating: 2 (Moderate negative)

Title: "Apple Is Pushing Back AI-Driven Siri Features to 2026, Report Says"
Content: Similar to the previous item, emphasizes delays in Siri enhancements.
Rating: 2 (Moderate negative)

Title: "Apple Says Some AI-Powered Enhancements to Siri to Be Delayed"
Content: Mentions delays in Siri updates amid slumping iPhone sales.
Rating: 2 (Moderate negative)

Title: "Sector Update: Tech Stocks Advance Late Afternoon"
Content: General update on tech stocks without specific focus on Apple.
Rating: 3 (Neutral)

Title: "Magnificent Seven Stocks: Nvidia Reverses Higher; Amazon, Meta, Tesla Slide"
Content: Discusses multiple tech stocks, with a brief mention of Apple’s mixed performance in 2025.
Rating: 3 (Neutral)

Title: "Apple says some AI improvements to Siri delayed to 2026"
Content: Reiterates delays in Siri's AI features, consistent with earlier reports.
Rating: 2 (Moderate negative)

Title: "Apple's First Foldable iPhone Could Launch in 2026 With Premium Pricing and AI Features"
Content: Speculates on a high-priced foldable iPhone with advanced features, suggesting innovation but no immediate impact.
Rating: 4 (Moderate positive)

Title: "Apple Inc. (AAPL): Jim Cramer Cautions – ‘I Could See the Stock Going Lower!’"
Content: Warns of potential downside for Apple stock, reflecting negativity.
Rating: 1 (Negative)

Title: "Apple Secures Local Content Certifications in Indonesia but Faces Additional Regulatory Hurdles"
Content: Notes progress in regulatory compliance but highlights challenges ahead.
Rating: 3 (Neutral)

Title: "Is Apple Inc. (AAPL) Among the Best Warren Buffett Stock to Buy Right Now?"
Content: Discusses Apple's position as a strong investment option under Warren Buffett’s philosophy, implying positivity.
Rating: 4 (Moderate positive)

Title: "The Mag 7 Stocks Are No Longer Magnificent. Where the Charts Say They Go Next."
Content: Broad commentary on tech stocks, including Apple, with no clear sentiment focus on Apple itself.
Rating: 3 (Neutral)

Title: "Broadcom Shares Jump as AI Growth Fuels Upbeat Forecast"
Content: Primarily about Broadcom, with indirect relevance to Apple as a client company.
Rating: 3 (Neutral)

Title: "Major Breakthrough: New Certificates Propel Apple iPhone 16 into Indonesia"
Content: Highlights progress in regulatory certification for iPhone sales in Indonesia, indicating positive developments for Apple’s market expansion.
Rating: 4 (Moderate positive)

Title: "In Rotten Market, Top Funds Harvest Apple Stock — And 10 Others"
Content: Indicates institutional confidence in Apple stock despite broader market challenges, reflecting positivity.
Rating: 4 (Moderate positive)

Title: "Apple's iPhone 16 Still on Hold in Indonesia Despite Key Certification Win"
Content: Acknowledges regulatory progress but emphasizes ongoing hurdles for product launch in Indonesia.
Rating: 3 (Neutral)

Title: "Cloud AI Today - Cirrascale Launches Inference Cloud For Scalable AI Integration"
Content: Not related to Apple stock specifically but mentions general AI developments in the industry.
Rating: 3 (Neutral)

Title: "38% of Berkshire Hathaway's Portfolio Is Invested in These 3 Unstoppable Dividend Stocks"
Content: General commentary on Warren Buffett’s portfolio with no specific focus on Apple stock sentiment.
Rating: 3 (Neutral)

Overall Sentiment Rating for Apple Stock
The individual ratings suggest a mix of sentiments:

Positive/Moderate Positive Ratings (5 or 4): 8 items

Neutral Ratings (3): 8 items

Negative/Moderate Negative Ratings (1 or 2): 4 items

The overall sentiment leans toward moderately positive for Apple stock, reflecting optimism tempered by some concerns over delays and market challenges.

Final Overall Sentiment Rating:
4 (Moderate positive)
```


## [Cointegration](#table-of-contents)

ChatGPT is not able to provide the stock with highest cointegration value with Apple stock:

```
The query about which S&P 500 stock has the highest cointegration value with Apple stock cannot be answered directly using the provided search results. Cointegration analysis requires statistical computations based on historical price data of Apple stock and other S&P 500 stocks. This involves time-series analysis to determine the long-term equilibrium relationship between Apple and other stocks.

If you have access to historical price data for Apple and other S&P 500 stocks, you can calculate cointegration values using statistical software or programming languages such as Python or R. The process typically involves:

Collecting Data: Obtain historical price data for Apple (AAPL) and other S&P 500 stocks over a specific period.

Stationarity Testing: Use tests like the Augmented Dickey-Fuller (ADF) test to ensure the time series are stationary.

Cointegration Testing: Apply tests such as the Engle-Granger two-step method or Johansen test to identify stocks with high cointegration values with Apple.

If you provide relevant price data or specify tools available, I can guide you further on how to perform this analysis.
```


## [Correlation](#table-of-contents)

ChatGPT 4-o stated Microsoft has the highest degree of correlation:

```
The stock most correlated with Apple (AAPL) among S&P 500 companies is Microsoft (MSFT). Historically, Apple and Microsoft have shown a high degree of correlation due to their positions as leading technology companies and their exposure to similar market dynamics, such as consumer electronics, cloud computing, and software services.

While specific numerical correlation values for S&P 500 stocks are not provided in the search results, Microsoft is widely recognized as one of Apple’s closest peers in terms of performance trends and market behavior. If you need precise correlation coefficients, you would need to calculate them using historical price data for Apple and other S&P 500 stocks over a specific period. Tools like Python, R, or financial platforms can help generate a correlation matrix for this purpose.
```


# [Compute Cointegration](#table-of-contents)

A number of issues with downloading OHLCV data via yfinance:

1. API misuse happen randomly dispite rate limtting and caching.
2. Inaccurate download of OHLCV data using yf.download(<ticker_list>) e.g.

- ['MMM', 'AOS', 'ABT', 'ABBV', 'ACN', 'ADBE', 'AMD', 'AES', 'AFL', 'A'] -> ABBV not downloaded but ABNB downloaded instead.
- ABNB downloaded twice.

As such, we have to resort to downloading individual ticker separately. Note that yfinance do not have OHLCV data for 'CTAS', 'BRK.B' and 'BF.B' tickers; therefore they are under the ignored_list.


In [4]:
co_integrate = CoIntegrate()
co_integrate.run()

YF.download() has changed argument auto_adjust default to True


[*********************100%***********************]  1 of 1 completed


    1) MMM    : 1305



[*********************100%***********************]  1 of 1 completed


    2) AOS    : 1305



[*********************100%***********************]  1 of 1 completed


    3) ABT    : 1305



[*********************100%***********************]  1 of 1 completed


    4) ABBV   : 1305



[*********************100%***********************]  1 of 1 completed


    5) ACN    : 1305



[*********************100%***********************]  1 of 1 completed


    6) ADBE   : 1305



[*********************100%***********************]  1 of 1 completed


    7) AMD    : 1305



[*********************100%***********************]  1 of 1 completed


    8) AES    : 1305



[*********************100%***********************]  1 of 1 completed


    9) AFL    : 1305



[*********************100%***********************]  1 of 1 completed


   10) A      : 1305



[*********************100%***********************]  1 of 1 completed


   11) APD    : 1305



[*********************100%***********************]  1 of 1 completed


   12) ABNB   : 1067



[*********************100%***********************]  1 of 1 completed


   13) AKAM   : 1305



[*********************100%***********************]  1 of 1 completed


   14) ALB    : 1305



[*********************100%***********************]  1 of 1 completed


   15) ARE    : 1305



[*********************100%***********************]  1 of 1 completed


   16) ALGN   : 1305



[*********************100%***********************]  1 of 1 completed


   17) ALLE   : 1305



[*********************100%***********************]  1 of 1 completed


   18) LNT    : 1305



[*********************100%***********************]  1 of 1 completed


   19) ALL    : 1305



[*********************100%***********************]  1 of 1 completed


   20) GOOGL  : 1305



[*********************100%***********************]  1 of 1 completed


   21) GOOG   : 1305



[*********************100%***********************]  1 of 1 completed


   22) MO     : 1305



[*********************100%***********************]  1 of 1 completed


   23) AMZN   : 1305



[*********************100%***********************]  1 of 1 completed


   24) AMCR   : 1305



[*********************100%***********************]  1 of 1 completed


   25) AEE    : 1305



[*********************100%***********************]  1 of 1 completed


   26) AEP    : 1305



[*********************100%***********************]  1 of 1 completed


   27) AXP    : 1305



[*********************100%***********************]  1 of 1 completed


   28) AIG    : 1305



[*********************100%***********************]  1 of 1 completed


   29) AMT    : 1305



[*********************100%***********************]  1 of 1 completed


   30) AWK    : 1305



[*********************100%***********************]  1 of 1 completed


   31) AMP    : 1305



[*********************100%***********************]  1 of 1 completed


   32) AME    : 1305



[*********************100%***********************]  1 of 1 completed


   33) AMGN   : 1305



[*********************100%***********************]  1 of 1 completed


   34) APH    : 1305



[*********************100%***********************]  1 of 1 completed


   35) ADI    : 1305



[*********************100%***********************]  1 of 1 completed


   36) ANSS   : 1305



[*********************100%***********************]  1 of 1 completed


   37) AON    : 1305



[*********************100%***********************]  1 of 1 completed


   38) APA    : 1305



[*********************100%***********************]  1 of 1 completed


   39) APO    : 1305



[*********************100%***********************]  1 of 1 completed


   40) AAPL   : 1305



[*********************100%***********************]  1 of 1 completed


   41) AMAT   : 1305



[*********************100%***********************]  1 of 1 completed


   42) APTV   : 1305



[*********************100%***********************]  1 of 1 completed


   43) ACGL   : 1305



[*********************100%***********************]  1 of 1 completed


   44) ADM    : 1305



[*********************100%***********************]  1 of 1 completed


   45) ANET   : 1305



[*********************100%***********************]  1 of 1 completed


   46) AJG    : 1305



[*********************100%***********************]  1 of 1 completed


   47) AIZ    : 1305



[*********************100%***********************]  1 of 1 completed


   48) T      : 1305



[*********************100%***********************]  1 of 1 completed


   49) ATO    : 1305



[*********************100%***********************]  1 of 1 completed


   50) ADSK   : 1305



[*********************100%***********************]  1 of 1 completed


   51) ADP    : 1305



[*********************100%***********************]  1 of 1 completed


   52) AZO    : 1305



[*********************100%***********************]  1 of 1 completed


   53) AVB    : 1305



[*********************100%***********************]  1 of 1 completed


   54) AVY    : 1305



[*********************100%***********************]  1 of 1 completed


   55) AXON   : 1305



[*********************100%***********************]  1 of 1 completed


   56) BKR    : 1305



[*********************100%***********************]  1 of 1 completed


   57) BALL   : 1305



[*********************100%***********************]  1 of 1 completed


   58) BAC    : 1305



[*********************100%***********************]  1 of 1 completed


   59) BAX    : 1305



[*********************100%***********************]  1 of 1 completed


   60) BDX    : 1305



[*********************100%***********************]  1 of 1 completed


   61) BBY    : 1305



[*********************100%***********************]  1 of 1 completed


   62) TECH   : 1305



[*********************100%***********************]  1 of 1 completed


   63) BIIB   : 1305



[*********************100%***********************]  1 of 1 completed


   64) BLK    : 1305



[*********************100%***********************]  1 of 1 completed


   65) BX     : 1305



[*********************100%***********************]  1 of 1 completed


   66) BK     : 1305



[*********************100%***********************]  1 of 1 completed


   67) BA     : 1305



[*********************100%***********************]  1 of 1 completed


   68) BKNG   : 1305



[*********************100%***********************]  1 of 1 completed


   69) BWA    : 1305



[*********************100%***********************]  1 of 1 completed


   70) BSX    : 1305



[*********************100%***********************]  1 of 1 completed


   71) BMY    : 1305



[*********************100%***********************]  1 of 1 completed


   72) AVGO   : 1305



[*********************100%***********************]  1 of 1 completed


   73) BR     : 1305



[*********************100%***********************]  1 of 1 completed


   74) BRO    : 1305



[*********************100%***********************]  1 of 1 completed


   75) BLDR   : 1305



[*********************100%***********************]  1 of 1 completed


   76) BG     : 1305



[*********************100%***********************]  1 of 1 completed


   77) BXP    : 1305



[*********************100%***********************]  1 of 1 completed


   78) CHRW   : 1305



[*********************100%***********************]  1 of 1 completed


   79) CDNS   : 1305



[*********************100%***********************]  1 of 1 completed


   80) CZR    : 1305



[*********************100%***********************]  1 of 1 completed


   81) CPT    : 1305



[*********************100%***********************]  1 of 1 completed


   82) CPB    : 1305



[*********************100%***********************]  1 of 1 completed


   83) COF    : 1305



[*********************100%***********************]  1 of 1 completed


   84) CAH    : 1305



[*********************100%***********************]  1 of 1 completed


   85) KMX    : 1305



[*********************100%***********************]  1 of 1 completed


   86) CCL    : 1305



[*********************100%***********************]  1 of 1 completed


   87) CARR   : 1252



[*********************100%***********************]  1 of 1 completed


   88) CAT    : 1305



[*********************100%***********************]  1 of 1 completed


   89) CBOE   : 1305



[*********************100%***********************]  1 of 1 completed


   90) CBRE   : 1305



[*********************100%***********************]  1 of 1 completed


   91) CDW    : 1305



[*********************100%***********************]  1 of 1 completed


   92) CE     : 1305



[*********************100%***********************]  1 of 1 completed


   93) COR    : 1305



[*********************100%***********************]  1 of 1 completed


   94) CNC    : 1305



[*********************100%***********************]  1 of 1 completed


   95) CNP    : 1305



[*********************100%***********************]  1 of 1 completed


   96) CF     : 1305



[*********************100%***********************]  1 of 1 completed


   97) CRL    : 1305



[*********************100%***********************]  1 of 1 completed


   98) SCHW   : 1305



[*********************100%***********************]  1 of 1 completed


   99) CHTR   : 1305



[*********************100%***********************]  1 of 1 completed


  100) CVX    : 1305



[*********************100%***********************]  1 of 1 completed


  101) CMG    : 1305



[*********************100%***********************]  1 of 1 completed


  102) CB     : 1305



[*********************100%***********************]  1 of 1 completed


  103) CHD    : 1305



[*********************100%***********************]  1 of 1 completed


  104) CI     : 1305



[*********************100%***********************]  1 of 1 completed


  105) CINF   : 1305



[*********************100%***********************]  1 of 1 completed

1 Failed download:
['CTAS']: YFPricesMissingError('possibly delisted; no price data found  (1d 2020-01-01 -> 2025-03-13)')


  106) CTAS   : 0



[*********************100%***********************]  1 of 1 completed


  107) CSCO   : 1305



[*********************100%***********************]  1 of 1 completed


  108) C      : 1305



[*********************100%***********************]  1 of 1 completed


  109) CFG    : 1305



[*********************100%***********************]  1 of 1 completed


  110) CLX    : 1305



[*********************100%***********************]  1 of 1 completed


  111) CME    : 1305



[*********************100%***********************]  1 of 1 completed


  112) CMS    : 1305



[*********************100%***********************]  1 of 1 completed


  113) KO     : 1305



[*********************100%***********************]  1 of 1 completed


  114) CTSH   : 1305



[*********************100%***********************]  1 of 1 completed


  115) CL     : 1305



[*********************100%***********************]  1 of 1 completed


  116) CMCSA  : 1305



[*********************100%***********************]  1 of 1 completed


  117) CAG    : 1305



[*********************100%***********************]  1 of 1 completed


  118) COP    : 1305



[*********************100%***********************]  1 of 1 completed


  119) ED     : 1305



[*********************100%***********************]  1 of 1 completed


  120) STZ    : 1305



[*********************100%***********************]  1 of 1 completed


  121) CEG    : 789



[*********************100%***********************]  1 of 1 completed


  122) COO    : 1305



[*********************100%***********************]  1 of 1 completed


  123) CPRT   : 1305



[*********************100%***********************]  1 of 1 completed


  124) GLW    : 1305



[*********************100%***********************]  1 of 1 completed


  125) CPAY   : 1305



[*********************100%***********************]  1 of 1 completed


  126) CTVA   : 1305



[*********************100%***********************]  1 of 1 completed


  127) CSGP   : 1305



[*********************100%***********************]  1 of 1 completed


  128) COST   : 1305



[*********************100%***********************]  1 of 1 completed


  129) CTRA   : 1305



[*********************100%***********************]  1 of 1 completed


  130) CRWD   : 1305



[*********************100%***********************]  1 of 1 completed


  131) CCI    : 1305



[*********************100%***********************]  1 of 1 completed


  132) CSX    : 1305



[*********************100%***********************]  1 of 1 completed


  133) CMI    : 1305



[*********************100%***********************]  1 of 1 completed


  134) CVS    : 1305



[*********************100%***********************]  1 of 1 completed


  135) DHR    : 1305



[*********************100%***********************]  1 of 1 completed


  136) DRI    : 1305



[*********************100%***********************]  1 of 1 completed


  137) DVA    : 1305



[*********************100%***********************]  1 of 1 completed


  138) DAY    : 1305



[*********************100%***********************]  1 of 1 completed


  139) DECK   : 1305



[*********************100%***********************]  1 of 1 completed


  140) DE     : 1305



[*********************100%***********************]  1 of 1 completed


  141) DELL   : 1305



[*********************100%***********************]  1 of 1 completed


  142) DAL    : 1305



[*********************100%***********************]  1 of 1 completed


  143) DVN    : 1305



[*********************100%***********************]  1 of 1 completed


  144) DXCM   : 1305



[*********************100%***********************]  1 of 1 completed


  145) FANG   : 1305



[*********************100%***********************]  1 of 1 completed


  146) DLR    : 1305



[*********************100%***********************]  1 of 1 completed


  147) DFS    : 1305



[*********************100%***********************]  1 of 1 completed


  148) DG     : 1305



[*********************100%***********************]  1 of 1 completed


  149) DLTR   : 1305



[*********************100%***********************]  1 of 1 completed


  150) D      : 1305



[*********************100%***********************]  1 of 1 completed


  151) DPZ    : 1305



[*********************100%***********************]  1 of 1 completed


  152) DOV    : 1305



[*********************100%***********************]  1 of 1 completed


  153) DOW    : 1305



[*********************100%***********************]  1 of 1 completed


  154) DHI    : 1305



[*********************100%***********************]  1 of 1 completed


  155) DTE    : 1305



[*********************100%***********************]  1 of 1 completed


  156) DUK    : 1305



[*********************100%***********************]  1 of 1 completed


  157) DD     : 1305



[*********************100%***********************]  1 of 1 completed


  158) EMN    : 1305



[*********************100%***********************]  1 of 1 completed


  159) ETN    : 1305



[*********************100%***********************]  1 of 1 completed


  160) EBAY   : 1305



[*********************100%***********************]  1 of 1 completed


  161) ECL    : 1305



[*********************100%***********************]  1 of 1 completed


  162) EIX    : 1305



[*********************100%***********************]  1 of 1 completed


  163) EW     : 1305



[*********************100%***********************]  1 of 1 completed


  164) EA     : 1305



[*********************100%***********************]  1 of 1 completed


  165) ELV    : 1305



[*********************100%***********************]  1 of 1 completed


  166) EMR    : 1305



[*********************100%***********************]  1 of 1 completed


  167) ENPH   : 1305



[*********************100%***********************]  1 of 1 completed


  168) ETR    : 1305



[*********************100%***********************]  1 of 1 completed


  169) EOG    : 1305



[*********************100%***********************]  1 of 1 completed


  170) EPAM   : 1305



[*********************100%***********************]  1 of 1 completed


  171) EQT    : 1305



[*********************100%***********************]  1 of 1 completed


  172) EFX    : 1305



[*********************100%***********************]  1 of 1 completed


  173) EQIX   : 1305



[*********************100%***********************]  1 of 1 completed


  174) EQR    : 1305



[*********************100%***********************]  1 of 1 completed


  175) ERIE   : 1305



[*********************100%***********************]  1 of 1 completed


  176) ESS    : 1305



[*********************100%***********************]  1 of 1 completed


  177) EL     : 1305



[*********************100%***********************]  1 of 1 completed


  178) EG     : 1305



[*********************100%***********************]  1 of 1 completed


  179) EVRG   : 1305



[*********************100%***********************]  1 of 1 completed


  180) ES     : 1305



[*********************100%***********************]  1 of 1 completed


  181) EXC    : 1305



[*********************100%***********************]  1 of 1 completed


  182) EXPE   : 1305



[*********************100%***********************]  1 of 1 completed


  183) EXPD   : 1305



[*********************100%***********************]  1 of 1 completed


  184) EXR    : 1305



[*********************100%***********************]  1 of 1 completed


  185) XOM    : 1305



[*********************100%***********************]  1 of 1 completed


  186) FFIV   : 1305



[*********************100%***********************]  1 of 1 completed


  187) FDS    : 1305



[*********************100%***********************]  1 of 1 completed


  188) FICO   : 1305



[*********************100%***********************]  1 of 1 completed


  189) FAST   : 1305



[*********************100%***********************]  1 of 1 completed


  190) FRT    : 1305



[*********************100%***********************]  1 of 1 completed


  191) FDX    : 1305



[*********************100%***********************]  1 of 1 completed


  192) FIS    : 1305



[*********************100%***********************]  1 of 1 completed


  193) FITB   : 1305



[*********************100%***********************]  1 of 1 completed


  194) FSLR   : 1305



[*********************100%***********************]  1 of 1 completed


  195) FE     : 1305



[*********************100%***********************]  1 of 1 completed


  196) FI     : 1305



[*********************100%***********************]  1 of 1 completed


  197) FMC    : 1305



[*********************100%***********************]  1 of 1 completed


  198) F      : 1305



[*********************100%***********************]  1 of 1 completed


  199) FTNT   : 1305



[*********************100%***********************]  1 of 1 completed


  200) FTV    : 1305



[*********************100%***********************]  1 of 1 completed


  201) FOXA   : 1305



[*********************100%***********************]  1 of 1 completed


  202) FOX    : 1305



[*********************100%***********************]  1 of 1 completed


  203) BEN    : 1305



[*********************100%***********************]  1 of 1 completed


  204) FCX    : 1305



[*********************100%***********************]  1 of 1 completed


  205) GRMN   : 1305



[*********************100%***********************]  1 of 1 completed


  206) IT     : 1305



[*********************100%***********************]  1 of 1 completed


  207) GE     : 1305



[*********************100%***********************]  1 of 1 completed


  208) GEHC   : 560



[*********************100%***********************]  1 of 1 completed


  209) GEV    : 238



[*********************100%***********************]  1 of 1 completed


  210) GEN    : 1305



[*********************100%***********************]  1 of 1 completed


  211) GNRC   : 1305



[*********************100%***********************]  1 of 1 completed


  212) GD     : 1305



[*********************100%***********************]  1 of 1 completed


  213) GIS    : 1305



[*********************100%***********************]  1 of 1 completed


  214) GM     : 1305



[*********************100%***********************]  1 of 1 completed


  215) GPC    : 1305



[*********************100%***********************]  1 of 1 completed


  216) GILD   : 1305



[*********************100%***********************]  1 of 1 completed


  217) GPN    : 1305



[*********************100%***********************]  1 of 1 completed


  218) GL     : 1305



[*********************100%***********************]  1 of 1 completed


  219) GDDY   : 1305



[*********************100%***********************]  1 of 1 completed


  220) GS     : 1305



[*********************100%***********************]  1 of 1 completed


  221) HAL    : 1305



[*********************100%***********************]  1 of 1 completed


  222) HIG    : 1305



[*********************100%***********************]  1 of 1 completed


  223) HAS    : 1305



[*********************100%***********************]  1 of 1 completed


  224) HCA    : 1305



[*********************100%***********************]  1 of 1 completed


  225) DOC    : 1305



[*********************100%***********************]  1 of 1 completed


  226) HSIC   : 1305



[*********************100%***********************]  1 of 1 completed


  227) HSY    : 1305



[*********************100%***********************]  1 of 1 completed


  228) HES    : 1305



[*********************100%***********************]  1 of 1 completed


  229) HPE    : 1305



[*********************100%***********************]  1 of 1 completed


  230) HLT    : 1305



[*********************100%***********************]  1 of 1 completed


  231) HOLX   : 1305



[*********************100%***********************]  1 of 1 completed


  232) HD     : 1305



[*********************100%***********************]  1 of 1 completed


  233) HON    : 1305



[*********************100%***********************]  1 of 1 completed


  234) HRL    : 1305



[*********************100%***********************]  1 of 1 completed


  235) HST    : 1305



[*********************100%***********************]  1 of 1 completed


  236) HWM    : 1305



[*********************100%***********************]  1 of 1 completed


  237) HPQ    : 1305



[*********************100%***********************]  1 of 1 completed


  238) HUBB   : 1305



[*********************100%***********************]  1 of 1 completed


  239) HUM    : 1305



[*********************100%***********************]  1 of 1 completed


  240) HBAN   : 1305



[*********************100%***********************]  1 of 1 completed


  241) HII    : 1305



[*********************100%***********************]  1 of 1 completed


  242) IBM    : 1305



[*********************100%***********************]  1 of 1 completed


  243) IEX    : 1305



[*********************100%***********************]  1 of 1 completed


  244) IDXX   : 1305



[*********************100%***********************]  1 of 1 completed


  245) ITW    : 1305



[*********************100%***********************]  1 of 1 completed


  246) INCY   : 1305



[*********************100%***********************]  1 of 1 completed


  247) IR     : 1305



[*********************100%***********************]  1 of 1 completed


  248) PODD   : 1305



[*********************100%***********************]  1 of 1 completed


  249) INTC   : 1305



[*********************100%***********************]  1 of 1 completed


  250) ICE    : 1305



[*********************100%***********************]  1 of 1 completed


  251) IFF    : 1305



[*********************100%***********************]  1 of 1 completed


  252) IP     : 1305



[*********************100%***********************]  1 of 1 completed


  253) IPG    : 1305



[*********************100%***********************]  1 of 1 completed


  254) INTU   : 1305



[*********************100%***********************]  1 of 1 completed


  255) ISRG   : 1305



[*********************100%***********************]  1 of 1 completed


  256) IVZ    : 1305



[*********************100%***********************]  1 of 1 completed


  257) INVH   : 1305



[*********************100%***********************]  1 of 1 completed


  258) IQV    : 1305



[*********************100%***********************]  1 of 1 completed


  259) IRM    : 1305



[*********************100%***********************]  1 of 1 completed


  260) JBHT   : 1305



[*********************100%***********************]  1 of 1 completed


  261) JBL    : 1305



[*********************100%***********************]  1 of 1 completed


  262) JKHY   : 1305



[*********************100%***********************]  1 of 1 completed


  263) J      : 1305



[*********************100%***********************]  1 of 1 completed


  264) JNJ    : 1305



[*********************100%***********************]  1 of 1 completed


  265) JCI    : 1305



[*********************100%***********************]  1 of 1 completed


  266) JPM    : 1305



[*********************100%***********************]  1 of 1 completed


  267) JNPR   : 1305



[*********************100%***********************]  1 of 1 completed


  268) K      : 1305



[*********************100%***********************]  1 of 1 completed


  269) KVUE   : 465



[*********************100%***********************]  1 of 1 completed


  270) KDP    : 1305



[*********************100%***********************]  1 of 1 completed


  271) KEY    : 1305



[*********************100%***********************]  1 of 1 completed


  272) KEYS   : 1305



[*********************100%***********************]  1 of 1 completed


  273) KMB    : 1305



[*********************100%***********************]  1 of 1 completed


  274) KIM    : 1305



[*********************100%***********************]  1 of 1 completed


  275) KMI    : 1305



[*********************100%***********************]  1 of 1 completed


  276) KKR    : 1305



[*********************100%***********************]  1 of 1 completed


  277) KLAC   : 1305



[*********************100%***********************]  1 of 1 completed


  278) KHC    : 1305



[*********************100%***********************]  1 of 1 completed


  279) KR     : 1305



[*********************100%***********************]  1 of 1 completed


  280) LHX    : 1305



[*********************100%***********************]  1 of 1 completed


  281) LH     : 1305



[*********************100%***********************]  1 of 1 completed


  282) LRCX   : 1305



[*********************100%***********************]  1 of 1 completed


  283) LW     : 1305



[*********************100%***********************]  1 of 1 completed


  284) LVS    : 1305



[*********************100%***********************]  1 of 1 completed


  285) LDOS   : 1305



[*********************100%***********************]  1 of 1 completed


  286) LEN    : 1305



[*********************100%***********************]  1 of 1 completed


  287) LII    : 1305



[*********************100%***********************]  1 of 1 completed


  288) LLY    : 1305



[*********************100%***********************]  1 of 1 completed


  289) LIN    : 1305



[*********************100%***********************]  1 of 1 completed


  290) LYV    : 1305



[*********************100%***********************]  1 of 1 completed


  291) LKQ    : 1305



[*********************100%***********************]  1 of 1 completed


  292) LMT    : 1305



[*********************100%***********************]  1 of 1 completed


  293) L      : 1305



[*********************100%***********************]  1 of 1 completed


  294) LOW    : 1305



[*********************100%***********************]  1 of 1 completed


  295) LULU   : 1305



[*********************100%***********************]  1 of 1 completed


  296) LYB    : 1305



[*********************100%***********************]  1 of 1 completed


  297) MTB    : 1305



[*********************100%***********************]  1 of 1 completed


  298) MPC    : 1305



[*********************100%***********************]  1 of 1 completed


  299) MKTX   : 1305



[*********************100%***********************]  1 of 1 completed


  300) MAR    : 1305



[*********************100%***********************]  1 of 1 completed


  301) MMC    : 1305



[*********************100%***********************]  1 of 1 completed


  302) MLM    : 1305



[*********************100%***********************]  1 of 1 completed


  303) MAS    : 1305



[*********************100%***********************]  1 of 1 completed


  304) MA     : 1305



[*********************100%***********************]  1 of 1 completed


  305) MTCH   : 1305



[*********************100%***********************]  1 of 1 completed


  306) MKC    : 1305



[*********************100%***********************]  1 of 1 completed


  307) MCD    : 1305



[*********************100%***********************]  1 of 1 completed


  308) MCK    : 1305



[*********************100%***********************]  1 of 1 completed


  309) MDT    : 1305



[*********************100%***********************]  1 of 1 completed


  310) MRK    : 1305



[*********************100%***********************]  1 of 1 completed


  311) META   : 1305



[*********************100%***********************]  1 of 1 completed


  312) MET    : 1305



[*********************100%***********************]  1 of 1 completed


  313) MTD    : 1305



[*********************100%***********************]  1 of 1 completed


  314) MGM    : 1305



[*********************100%***********************]  1 of 1 completed


  315) MCHP   : 1305



[*********************100%***********************]  1 of 1 completed


  316) MU     : 1305



[*********************100%***********************]  1 of 1 completed


  317) MSFT   : 1305



[*********************100%***********************]  1 of 1 completed


  318) MAA    : 1305



[*********************100%***********************]  1 of 1 completed


  319) MRNA   : 1305



[*********************100%***********************]  1 of 1 completed


  320) MHK    : 1305



[*********************100%***********************]  1 of 1 completed


  321) MOH    : 1305



[*********************100%***********************]  1 of 1 completed


  322) TAP    : 1305



[*********************100%***********************]  1 of 1 completed


  323) MDLZ   : 1305



[*********************100%***********************]  1 of 1 completed


  324) MPWR   : 1305



[*********************100%***********************]  1 of 1 completed


  325) MNST   : 1305



[*********************100%***********************]  1 of 1 completed


  326) MCO    : 1305



[*********************100%***********************]  1 of 1 completed


  327) MS     : 1305



[*********************100%***********************]  1 of 1 completed


  328) MOS    : 1305



[*********************100%***********************]  1 of 1 completed


  329) MSI    : 1305



[*********************100%***********************]  1 of 1 completed


  330) MSCI   : 1305



[*********************100%***********************]  1 of 1 completed


  331) NDAQ   : 1305



[*********************100%***********************]  1 of 1 completed


  332) NTAP   : 1305



[*********************100%***********************]  1 of 1 completed


  333) NFLX   : 1305



[*********************100%***********************]  1 of 1 completed


  334) NEM    : 1305



[*********************100%***********************]  1 of 1 completed


  335) NWSA   : 1305



[*********************100%***********************]  1 of 1 completed


  336) NWS    : 1305



[*********************100%***********************]  1 of 1 completed


  337) NEE    : 1305



[*********************100%***********************]  1 of 1 completed


  338) NKE    : 1305



[*********************100%***********************]  1 of 1 completed


  339) NI     : 1305



[*********************100%***********************]  1 of 1 completed


  340) NDSN   : 1305



[*********************100%***********************]  1 of 1 completed


  341) NSC    : 1305



[*********************100%***********************]  1 of 1 completed


  342) NTRS   : 1305



[*********************100%***********************]  1 of 1 completed


  343) NOC    : 1305



[*********************100%***********************]  1 of 1 completed


  344) NCLH   : 1305



[*********************100%***********************]  1 of 1 completed


  345) NRG    : 1305



[*********************100%***********************]  1 of 1 completed


  346) NUE    : 1305



[*********************100%***********************]  1 of 1 completed


  347) NVDA   : 1305



[*********************100%***********************]  1 of 1 completed


  348) NVR    : 1305



[*********************100%***********************]  1 of 1 completed


  349) NXPI   : 1305



[*********************100%***********************]  1 of 1 completed


  350) ORLY   : 1305



[*********************100%***********************]  1 of 1 completed


  351) OXY    : 1305



[*********************100%***********************]  1 of 1 completed


  352) ODFL   : 1305



[*********************100%***********************]  1 of 1 completed


  353) OMC    : 1305



[*********************100%***********************]  1 of 1 completed


  354) ON     : 1305



[*********************100%***********************]  1 of 1 completed


  355) OKE    : 1305



[*********************100%***********************]  1 of 1 completed


  356) ORCL   : 1305



[*********************100%***********************]  1 of 1 completed


  357) OTIS   : 1252



[*********************100%***********************]  1 of 1 completed


  358) PCAR   : 1305



[*********************100%***********************]  1 of 1 completed


  359) PKG    : 1305



[*********************100%***********************]  1 of 1 completed


  360) PLTR   : 1117



[*********************100%***********************]  1 of 1 completed


  361) PANW   : 1305



[*********************100%***********************]  1 of 1 completed


  362) PARA   : 1305



[*********************100%***********************]  1 of 1 completed


  363) PH     : 1305



[*********************100%***********************]  1 of 1 completed


  364) PAYX   : 1305



[*********************100%***********************]  1 of 1 completed


  365) PAYC   : 1305



[*********************100%***********************]  1 of 1 completed


  366) PYPL   : 1305



[*********************100%***********************]  1 of 1 completed


  367) PNR    : 1305



[*********************100%***********************]  1 of 1 completed


  368) PEP    : 1305



[*********************100%***********************]  1 of 1 completed


  369) PFE    : 1305



[*********************100%***********************]  1 of 1 completed


  370) PCG    : 1305



[*********************100%***********************]  1 of 1 completed


  371) PM     : 1305



[*********************100%***********************]  1 of 1 completed


  372) PSX    : 1305



[*********************100%***********************]  1 of 1 completed


  373) PNW    : 1305



[*********************100%***********************]  1 of 1 completed


  374) PNC    : 1305



[*********************100%***********************]  1 of 1 completed


  375) POOL   : 1305



[*********************100%***********************]  1 of 1 completed


  376) PPG    : 1305



[*********************100%***********************]  1 of 1 completed


  377) PPL    : 1305



[*********************100%***********************]  1 of 1 completed


  378) PFG    : 1305



[*********************100%***********************]  1 of 1 completed


  379) PG     : 1305



[*********************100%***********************]  1 of 1 completed


  380) PGR    : 1305



[*********************100%***********************]  1 of 1 completed


  381) PLD    : 1305



[*********************100%***********************]  1 of 1 completed


  382) PRU    : 1305



[*********************100%***********************]  1 of 1 completed


  383) PEG    : 1305



[*********************100%***********************]  1 of 1 completed


  384) PTC    : 1305



[*********************100%***********************]  1 of 1 completed


  385) PSA    : 1305



[*********************100%***********************]  1 of 1 completed


  386) PHM    : 1305



[*********************100%***********************]  1 of 1 completed


  387) PWR    : 1305



[*********************100%***********************]  1 of 1 completed


  388) QCOM   : 1305



[*********************100%***********************]  1 of 1 completed


  389) DGX    : 1305



[*********************100%***********************]  1 of 1 completed


  390) RL     : 1305



[*********************100%***********************]  1 of 1 completed


  391) RJF    : 1305



[*********************100%***********************]  1 of 1 completed


  392) RTX    : 1305



[*********************100%***********************]  1 of 1 completed


  393) O      : 1305



[*********************100%***********************]  1 of 1 completed


  394) REG    : 1305



[*********************100%***********************]  1 of 1 completed


  395) REGN   : 1305



[*********************100%***********************]  1 of 1 completed


  396) RF     : 1305



[*********************100%***********************]  1 of 1 completed


  397) RSG    : 1305



[*********************100%***********************]  1 of 1 completed


  398) RMD    : 1305



[*********************100%***********************]  1 of 1 completed


  399) RVTY   : 1305



[*********************100%***********************]  1 of 1 completed


  400) ROK    : 1305



[*********************100%***********************]  1 of 1 completed


  401) ROL    : 1305



[*********************100%***********************]  1 of 1 completed


  402) ROP    : 1305



[*********************100%***********************]  1 of 1 completed


  403) ROST   : 1305



[*********************100%***********************]  1 of 1 completed


  404) RCL    : 1305



[*********************100%***********************]  1 of 1 completed


  405) SPGI   : 1305



[*********************100%***********************]  1 of 1 completed


  406) CRM    : 1305



[*********************100%***********************]  1 of 1 completed


  407) SBAC   : 1305



[*********************100%***********************]  1 of 1 completed


  408) SLB    : 1305



[*********************100%***********************]  1 of 1 completed


  409) STX    : 1305



[*********************100%***********************]  1 of 1 completed


  410) SRE    : 1305



[*********************100%***********************]  1 of 1 completed


  411) NOW    : 1305



[*********************100%***********************]  1 of 1 completed


  412) SHW    : 1305



[*********************100%***********************]  1 of 1 completed


  413) SPG    : 1305



[*********************100%***********************]  1 of 1 completed


  414) SWKS   : 1305



[*********************100%***********************]  1 of 1 completed


  415) SJM    : 1305



[*********************100%***********************]  1 of 1 completed


  416) SW     : 171



[*********************100%***********************]  1 of 1 completed


  417) SNA    : 1305



[*********************100%***********************]  1 of 1 completed


  418) SOLV   : 238



[*********************100%***********************]  1 of 1 completed


  419) SO     : 1305



[*********************100%***********************]  1 of 1 completed


  420) LUV    : 1305



[*********************100%***********************]  1 of 1 completed


  421) SWK    : 1305



[*********************100%***********************]  1 of 1 completed


  422) SBUX   : 1305



[*********************100%***********************]  1 of 1 completed


  423) STT    : 1305



[*********************100%***********************]  1 of 1 completed


  424) STLD   : 1305



[*********************100%***********************]  1 of 1 completed


  425) STE    : 1305



[*********************100%***********************]  1 of 1 completed


  426) SYK    : 1305



[*********************100%***********************]  1 of 1 completed


  427) SMCI   : 1305



[*********************100%***********************]  1 of 1 completed


  428) SYF    : 1305



[*********************100%***********************]  1 of 1 completed


  429) SNPS   : 1305



[*********************100%***********************]  1 of 1 completed


  430) SYY    : 1305



[*********************100%***********************]  1 of 1 completed


  431) TMUS   : 1305



[*********************100%***********************]  1 of 1 completed


  432) TROW   : 1305



[*********************100%***********************]  1 of 1 completed


  433) TTWO   : 1305



[*********************100%***********************]  1 of 1 completed


  434) TPR    : 1305



[*********************100%***********************]  1 of 1 completed


  435) TRGP   : 1305



[*********************100%***********************]  1 of 1 completed


  436) TGT    : 1305



[*********************100%***********************]  1 of 1 completed


  437) TEL    : 1305



[*********************100%***********************]  1 of 1 completed


  438) TDY    : 1305



[*********************100%***********************]  1 of 1 completed


  439) TFX    : 1305



[*********************100%***********************]  1 of 1 completed


  440) TER    : 1305



[*********************100%***********************]  1 of 1 completed


  441) TSLA   : 1305



[*********************100%***********************]  1 of 1 completed


  442) TXN    : 1305



[*********************100%***********************]  1 of 1 completed


  443) TPL    : 1305



[*********************100%***********************]  1 of 1 completed


  444) TXT    : 1305



[*********************100%***********************]  1 of 1 completed


  445) TMO    : 1305



[*********************100%***********************]  1 of 1 completed


  446) TJX    : 1305



[*********************100%***********************]  1 of 1 completed


  447) TSCO   : 1305



[*********************100%***********************]  1 of 1 completed


  448) TT     : 1305



[*********************100%***********************]  1 of 1 completed


  449) TDG    : 1305



[*********************100%***********************]  1 of 1 completed


  450) TRV    : 1305



[*********************100%***********************]  1 of 1 completed


  451) TRMB   : 1305



[*********************100%***********************]  1 of 1 completed


  452) TFC    : 1305



[*********************100%***********************]  1 of 1 completed


  453) TYL    : 1305



[*********************100%***********************]  1 of 1 completed


  454) TSN    : 1305



[*********************100%***********************]  1 of 1 completed


  455) USB    : 1305



[*********************100%***********************]  1 of 1 completed


  456) UBER   : 1305



[*********************100%***********************]  1 of 1 completed


  457) UDR    : 1305



[*********************100%***********************]  1 of 1 completed


  458) ULTA   : 1305



[*********************100%***********************]  1 of 1 completed


  459) UNP    : 1305



[*********************100%***********************]  1 of 1 completed


  460) UAL    : 1305



[*********************100%***********************]  1 of 1 completed


  461) UPS    : 1305



[*********************100%***********************]  1 of 1 completed


  462) URI    : 1305



[*********************100%***********************]  1 of 1 completed


  463) UNH    : 1305



[*********************100%***********************]  1 of 1 completed


  464) UHS    : 1305



[*********************100%***********************]  1 of 1 completed


  465) VLO    : 1305



[*********************100%***********************]  1 of 1 completed


  466) VTR    : 1305



[*********************100%***********************]  1 of 1 completed


  467) VLTO   : 360



[*********************100%***********************]  1 of 1 completed


  468) VRSN   : 1305



[*********************100%***********************]  1 of 1 completed


  469) VRSK   : 1305



[*********************100%***********************]  1 of 1 completed


  470) VZ     : 1305



[*********************100%***********************]  1 of 1 completed


  471) VRTX   : 1305



[*********************100%***********************]  1 of 1 completed


  472) VTRS   : 1305



[*********************100%***********************]  1 of 1 completed


  473) VICI   : 1305



[*********************100%***********************]  1 of 1 completed


  474) V      : 1305



[*********************100%***********************]  1 of 1 completed


  475) VST    : 1305



[*********************100%***********************]  1 of 1 completed


  476) VMC    : 1305



[*********************100%***********************]  1 of 1 completed


  477) WRB    : 1305



[*********************100%***********************]  1 of 1 completed


  478) GWW    : 1305



[*********************100%***********************]  1 of 1 completed


  479) WAB    : 1305



[*********************100%***********************]  1 of 1 completed


  480) WBA    : 1305



[*********************100%***********************]  1 of 1 completed


  481) WMT    : 1305



[*********************100%***********************]  1 of 1 completed


  482) DIS    : 1305



[*********************100%***********************]  1 of 1 completed


  483) WBD    : 1305



[*********************100%***********************]  1 of 1 completed


  484) WM     : 1305



[*********************100%***********************]  1 of 1 completed


  485) WAT    : 1305



[*********************100%***********************]  1 of 1 completed


  486) WEC    : 1305



[*********************100%***********************]  1 of 1 completed


  487) WFC    : 1305



[*********************100%***********************]  1 of 1 completed


  488) WELL   : 1305



[*********************100%***********************]  1 of 1 completed


  489) WST    : 1305



[*********************100%***********************]  1 of 1 completed


  490) WDC    : 1305



[*********************100%***********************]  1 of 1 completed


  491) WY     : 1305



[*********************100%***********************]  1 of 1 completed


  492) WMB    : 1305



[*********************100%***********************]  1 of 1 completed


  493) WTW    : 1305



[*********************100%***********************]  1 of 1 completed


  494) WDAY   : 1305



[*********************100%***********************]  1 of 1 completed


  495) WYNN   : 1305



[*********************100%***********************]  1 of 1 completed


  496) XEL    : 1305



[*********************100%***********************]  1 of 1 completed


  497) XYL    : 1305



[*********************100%***********************]  1 of 1 completed


  498) YUM    : 1305



[*********************100%***********************]  1 of 1 completed


  499) ZBRA   : 1305



[*********************100%***********************]  1 of 1 completed


  500) ZBH    : 1305



[*********************100%***********************]  1 of 1 completed

  501) ZTS    : 1305



In [ ]:
# 'CTAS' OHLCV dta is not available in yfinance
yf.download("CTAS", start="2020-01-01")

[*********************100%***********************]  1 of 1 completed

1 Failed download:
['CTAS']: YFPricesMissingError('possibly delisted; no price data found  (1d 2020-01-01 -> 2025-03-13)')


Empty DataFrame
Columns: [(Adj Close, CTAS), (Close, CTAS), (High, CTAS), (Low, CTAS), (Open, CTAS), (Volume, CTAS)]
Index: []

In [4]:
# Cointegration stocks with 'AAPL' ticker
df_28 = utils.get_coint_ticker("AAPL", "2025-03-28")
df_28

ticker1 ticker2    pvalue   coint_t  percent_1  percent_5  percent_10  \
18170    AAPL      PG  0.000914 -4.576653  -3.905181  -3.340999   -3.047828   
18039    AAPL    INCY  0.005623 -4.073665  -3.905181  -3.340999   -3.047828   
18121    AAPL     MSI  0.006085 -4.050084  -3.905181  -3.340999   -3.047828   
18208    AAPL      SO  0.006984 -4.008461  -3.905181  -3.340999   -3.047828   
18148    AAPL    ORCL  0.007844 -3.972982  -3.905181  -3.340999   -3.047828   
17845    AAPL    ANET  0.008812 -3.937046  -3.905181  -3.340999   -3.047828   
18022    AAPL     HPE  0.009419 -3.916261  -3.905181  -3.340999   -3.047828   
18152    AAPL    PANW  0.010237 -3.890069  -3.905181  -3.340999   -3.047828   
18178    AAPL     PWR  0.010473 -3.882867  -3.905181  -3.340999   -3.047828   
17944    AAPL      DG  0.011050 -3.865832  -3.905181  -3.340999   -3.047828   
18023    AAPL     HLT  0.012626 -3.823058  -3.905181  -3.340999   -3.047828   
17856    AAPL     BKR  0.014014 -3.789144  -3.905181  -3.340999   -3.047828   
18085    AAPL       L  0.014618 -3.775306  -3.905181  -3.340999   -3.047828   
18147    AAPL     OKE  0.014662 -3.774318  -3.905181  -3.340999   -3.047828   
18027    AAPL     HRL  0.015148 -3.763589  -3.905181  -3.340999   -3.047828   
18174    AAPL     PEG  0.015544 -3.755050  -3.905181  -3.340999   -3.047828   
18052    AAPL     IRM  0.015839 -3.748830  -3.905181  -3.340999   -3.047828   
17924    AAPL    COST  0.016530 -3.734621  -3.905181  -3.340999   -3.047828   
17888    AAPL    CBOE  0.016910 -3.727030  -3.905181  -3.340999   -3.047828   
17916    AAPL      ED  0.019577 -3.677586  -3.905181  -3.340999   -3.047828   
18092    AAPL     MAR  0.020426 -3.663081  -3.905181  -3.340999   -3.047828   
17871    AAPL    AVGO  0.022105 -3.635877  -3.905181  -3.340999   -3.047828   
17900    AAPL      CB  0.022818 -3.624868  -3.905181  -3.340999   -3.047828   
18151    AAPL     PKG  0.023483 -3.614868  -3.905181  -3.340999   -3.047828   
17846    AAPL     AJG  0.025093 -3.591620  -3.905181  -3.340999   -3.047828   
18093    AAPL     MMC  0.025353 -3.587980  -3.905181  -3.340999   -3.047828   
18091    AAPL    MKTX  0.026650 -3.570330  -3.905181  -3.340999   -3.047828   
18224    AAPL    TRGP  0.028001 -3.552693  -3.905181  -3.340999   -3.047828   
18267    AAPL     WAB  0.029928 -3.528774  -3.905181  -3.340999   -3.047828   
4862      APD    AAPL  0.032147 -3.502809  -3.905181  -3.340999   -3.047828   
17873    AAPL     BRO  0.033052 -3.492653  -3.905181  -3.340999   -3.047828   
18139    AAPL    NVDA  0.033088 -3.492255  -3.905181  -3.340999   -3.047828   
18236    AAPL    TSCO  0.033363 -3.489220  -3.905181  -3.340999   -3.047828   
18188    AAPL     RSG  0.033761 -3.484864  -3.905181  -3.340999   -3.047828   
18264    AAPL     VMC  0.034732 -3.474417  -3.905181  -3.340999   -3.047828   
18183    AAPL     RTX  0.034835 -3.473316  -3.905181  -3.340999   -3.047828   
18272    AAPL      WM  0.035288 -3.468542  -3.905181  -3.340999   -3.047828   
17887    AAPL     CAT  0.036236 -3.458708  -3.905181  -3.340999   -3.047828   
18275    AAPL     WFC  0.037567 -3.445249  -3.905181  -3.340999   -3.047828   
18071    AAPL      KR  0.039127 -3.429988  -3.905181  -3.340999   -3.047828   
18164    AAPL     PNW  0.039300 -3.428333  -3.905181  -3.340999   -3.047828   
17843    AAPL    ACGL  0.040337 -3.418497  -3.905181  -3.340999   -3.047828   
18015    AAPL     HIG  0.043868 -3.386541  -3.905181  -3.340999   -3.047828   
10520    AMCR    AAPL  0.045204 -3.375000  -3.905181  -3.340999   -3.047828   
18080    AAPL     LLY  0.045516 -3.372351  -3.905181  -3.340999   -3.047828   
17891    AAPL     COR  0.045592 -3.371706  -3.905181  -3.340999   -3.047828   
17955    AAPL     ETN  0.045951 -3.368677  -3.905181  -3.340999   -3.047828   
17973    AAPL      EL  0.046249 -3.366175  -3.905181  -3.340999   -3.047828   
18131    AAPL      NI  0.046499 -3.364089  -3.905181  -3.340999   -3.047828   
18100    AAPL     MCK  0.046922 -3.360574  -3.905181  -3

In [3]:
from scipy import stats

res = stats.spearmanr([1, 2, 3, 4, 5], [5, 6, 7, 8, 7])
res.statistic, res.pvalue

(np.float64(0.8207826816681233), np.float64(0.08858700531354381))

In [4]:
res = stats.pearsonr([1, 2, 3, 4, 5], [5, 6, 7, 8, 7])
res.statistic, res.pvalue

(np.float64(0.8320502943378436), np.float64(0.0805095732984986))

In [5]:
res = stats.kendalltau([1, 2, 3, 4, 5], [5, 6, 7, 8, 7])
res.statistic, res.pvalue

(np.float64(0.7378647873726218), np.float64(0.07697417298126674))

## Analyze AAPL


In [5]:
analyze_coint = AnalyzeCoint()
df_coint_analysis = analyze_coint("AAPL")
df_coint_analysis

date1  period1                                    top_20_tickers1  \
0   2025-03-13        1  VLO, PSX, CSGP, COP, TJX, LII, MOS, MPC, EQIX,...   
1   2025-03-13        1  VLO, PSX, CSGP, COP, TJX, LII, MOS, MPC, EQIX,...   
2   2025-03-13        1  VLO, PSX, CSGP, COP, TJX, LII, MOS, MPC, EQIX,...   
3   2025-03-13        1  VLO, PSX, CSGP, COP, TJX, LII, MOS, MPC, EQIX,...   
4   2025-03-13        1  VLO, PSX, CSGP, COP, TJX, LII, MOS, MPC, EQIX,...   
5   2025-03-13        1  VLO, PSX, CSGP, COP, TJX, LII, MOS, MPC, EQIX,...   
6   2025-03-13        1  VLO, PSX, CSGP, COP, TJX, LII, MOS, MPC, EQIX,...   
7   2025-03-13        1  VLO, PSX, CSGP, COP, TJX, LII, MOS, MPC, EQIX,...   
8   2025-03-13        3  DVN, APA, OXY, VRSK, TYL, ORCL, SPGI, EL, LKQ,...   
9   2025-03-13        3  DVN, APA, OXY, VRSK, TYL, ORCL, SPGI, EL, LKQ,...   
10  2025-03-13        3  DVN, APA, OXY, VRSK, TYL, ORCL, SPGI, EL, LKQ,...   
11  2025-03-13        3  DVN, APA, OXY, VRSK, TYL, ORCL, SPGI, EL, LKQ,...   
12  2025-03-13        3  DVN, APA, OXY, VRSK, TYL, ORCL, SPGI, EL, LKQ,...   
13  2025-03-13        3  DVN, APA, OXY, VRSK, TYL, ORCL, SPGI, EL, LKQ,...   
14  2025-03-13        3  DVN, APA, OXY, VRSK, TYL, ORCL, SPGI, EL, LKQ,...   
15  2025-03-13        5  PG, OKE, BKR, HPE, MAR, L, PEG, WFC, SPG, PPL,...   
16  2025-03-13        5  PG, OKE, BKR, HPE, MAR, L, PEG, WFC, SPG, PPL,...   
17  2025-03-13        5  PG, OKE, BKR, HPE, MAR, L, PEG, WFC, SPG, PPL,...   
18  2025-03-13        5  PG, OKE, BKR, HPE, MAR, L, PEG, WFC, SPG, PPL,...   
19  2025-03-13        5  PG, OKE, BKR, HPE, MAR, L, PEG, WFC, SPG, PPL,...   
20  2025-03-13        5  PG, OKE, BKR, HPE, MAR, L, PEG, WFC, SPG, PPL,...   
21  2025-03-26        1  LII, MOS, PSX, TJX, VLO, CSGP, ZBH, GEN, NVDA,...   
22  2025-03-26        1  LII, MOS, PSX, TJX, VLO, CSGP, ZBH, GEN, NVDA,...   
23  2025-03-26        1  LII, MOS, PSX, TJX, VLO, CSGP, ZBH, GEN, NVDA,...   
24  2025-03-26        1  LII, MOS, PSX, TJX, VLO, CSGP, ZBH, GEN, NVDA,...   
25  2025-03-26        1  LII, MOS, PSX, TJX, VLO, CSGP, ZBH, GEN, NVDA,...   
26  2025-03-26        3  APA, VRSK, ORCL, DVN, EL, WBA, TYL, OXY, ARE, ...   
27  2025-03-26        3  APA, VRSK, ORCL, DVN, EL, WBA, TYL, OXY, ARE, ...   
28  2025-03-26        3  APA, VRSK, ORCL, DVN, EL, WBA, TYL, OXY, ARE, ...   
29  2025-03-26        3  APA, VRSK, ORCL, DVN, EL, WBA, TYL, OXY, ARE, ...   
30  2025-03-26        5  PG, OKE, MSI, HRL, HPE, PANW, INCY, DG, ANET, ...   
31  2025-03-26        5  PG, OKE, MSI, HRL, HPE, PANW, INCY, DG, ANET, ...   
32  2025-03-26        5  PG, OKE, MSI, HRL, HPE, PANW, INCY, DG, ANET, ...   
33  2025-03-28        1  MOS, TJX, LII, VLO, PSX, CSGP, KMX, GEN, ZBH, ...   
34  2025-03-28        1  MOS, TJX, LII, VLO, PSX, CSGP, KMX, GEN, ZBH, ...   
35  2025-03-28        3  APA, EL, VRSK, WBA, ARE, DVN, ORCL, OXY, TYL, ...   

         date2  period2                                    top_20_tickers2  \
0   2025-03-13        3  DVN, APA, OXY, VRSK, TYL, ORCL, SPGI, EL, LKQ,...   
1   2025-03-13        5  PG, OKE, BKR, HPE, MAR, L, PEG, WFC, SPG, PPL,...   
2   2025-03-26        1  LII, MOS, PSX, TJX, VLO, CSGP, ZBH, GEN, NVDA,...   
3   2025-03-26        3  APA, VRSK, ORCL, DVN, EL, WBA, TYL, OXY, ARE, ...   
4   2025-03-26        5  PG, OKE, MSI, HRL, HPE, PANW, INCY, DG, ANET, ...   
5   2025-03-28        1  MOS, TJX, LII, VLO, PSX, CSGP, KMX, GEN, ZBH, ...   
6   2025-03-28        3  APA, EL, VRSK, WBA, ARE, DVN, ORCL, OXY, TYL, ...   
7   2025-03-28        5  PG, INCY, MSI, SO, ORCL, ANET, HPE, PANW, PWR,...   
8   2025-03-13        5  PG, OKE, BKR, HPE, MAR, L, PEG, WFC, SPG, PPL,...   
9   2025-03-26        1  LII, MOS, PSX, TJX, VLO, CSGP, ZBH, GEN, NVDA,...   
10  2025-03-26        3  APA, VRSK, ORCL, DVN, EL, WBA, TYL, OXY, ARE, ...   
11  2025-03-26        5  PG, OKE, MSI, HRL, HPE, PANW, INCY, DG, ANET, ...   
12  2025-03-28        1  MOS, TJX, LII, VLO, PSX, CSGP, KMX, GEN, ZBH, ...   
13  2025-03-28        3  APA, 

In [18]:
df_coint_analysis.loc[
    df_coint_analysis["period1"] == df_coint_analysis["period2"], :
].sort_values(by=["date1", "period1"])

date1  period1                                    top_20_tickers1  \
28  2025-03-13        1  VLO, PSX, CSGP, COP, TJX, LII, MOS, MPC, EQIX,...   
23  2025-03-13        3  DVN, APA, OXY, VRSK, TYL, ORCL, SPGI, EL, LKQ,...   
32  2025-03-13        5  PG, OKE, BKR, HPE, MAR, L, PEG, WFC, SPG, PPL,...   
10  2025-03-26        1  LII, MOS, PSX, TJX, VLO, CSGP, ZBH, GEN, NVDA,...   
13  2025-03-26        1  LII, MOS, PSX, TJX, VLO, CSGP, ZBH, GEN, NVDA,...   
2   2025-03-26        3  APA, VRSK, ORCL, DVN, EL, WBA, TYL, OXY, ARE, ...   
5   2025-03-26        3  APA, VRSK, ORCL, DVN, EL, WBA, TYL, OXY, ARE, ...   
17  2025-03-26        5  PG, OKE, MSI, HRL, HPE, PANW, INCY, DG, ANET, ...   
20  2025-03-26        5  PG, OKE, MSI, HRL, HPE, PANW, INCY, DG, ANET, ...   

         date2  period2                                    top_20_tickers2  \
28  2025-03-28        1  MOS, TJX, LII, VLO, PSX, CSGP, KMX, GEN, ZBH, ...   
23  2025-03-28        3  APA, EL, VRSK, WBA, ARE, DVN, ORCL, OXY, TYL, ...   
32  2025-03-28        5  PG, INCY, MSI, SO, ORCL, ANET, HPE, PANW, PWR,...   
10  2025-03-13        1  VLO, PSX, CSGP, COP, TJX, LII, MOS, MPC, EQIX,...   
13  2025-03-28        1  MOS, TJX, LII, VLO, PSX, CSGP, KMX, GEN, ZBH, ...   
2   2025-03-13        3  DVN, APA, OXY, VRSK, TYL, ORCL, SPGI, EL, LKQ,...   
5   2025-03-28        3  APA, EL, VRSK, WBA, ARE, DVN, ORCL, OXY, TYL, ...   
17  2025-03-13        5  PG, OKE, BKR, HPE, MAR, L, PEG, WFC, SPG, PPL,...   
20  2025-03-28        5  PG, INCY, MSI, SO, ORCL, ANET, HPE, PANW, PWR,...   

    num_common                                     common_tickers  
28           7                 VLO, PSX, CSGP, TJX, LII, MOS, ZBH  
23          15  DVN, APA, OXY, VRSK, TYL, ORCL, SPGI, EL, MCO,...  
32           8                 PG, OKE, BKR, HPE, L, PEG, HLT, SO  
10           8            LII, MOS, PSX, TJX, VLO, CSGP, ZBH, COP  
13          10  LII, MOS, PSX, TJX, VLO, CSGP, ZBH, GEN, NVDA,...  
2           15  APA, VRSK, ORCL, DVN, EL, WBA, TYL, OXY, ANET,...  
5           18  APA, VRSK, ORCL, DVN, EL, WBA, TYL, OXY, ARE, ...  
17           9            PG, OKE, HPE, HLT, SO, BKR, L, MAR, PEG  
20          19  PG, OKE, MSI, HRL, HPE, PANW, INCY, DG, ANET, ...

# [Sentiment](#table-of-contents)

- Capture publication date, publisher (e.g. Bloomberg), news title and news content via yfinance news filtered by ticker e.g. https://finance.yahoo.com/quote/AAPL/news/
- Use different HuggingFace FinBERT on news title only, news content only and both news title and content:

| FinBERT                                                                                    | Model Card                            | Column      | Description                                                                                                                                                      |
| ------------------------------------------------------------------------------------------ | ------------------------------------- | ----------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| [Prosus AI](https://huggingface.co/ProsusAI/finbert)                                       | ProsusAI/finbert                      | prosusai    | &bull; Pre-trained NLP model to analyze sentiment of financial text. <br>&bull; Used Financial PhraseBank by Malo et al. (2014) is used for fine-tuning.         |
| [finbert-tone](https://huggingface.co/yiyanghkust/finbert-tone)                            | yiyanghkust/finbert-tone              | yiyanghkust | &bull; Pre-trained on corporate reports, earning calls, and analyst reports. <br>&bull; Fined-tuned on 10,000 manually annotated sentences from analyst reports. |
| [FinBERT-FOMC](https://huggingface.co/ZiweiChen/FinBERT-FOMC)                              | ZiweiChen/FinBERT-FOMC                | ziweichen   | &bull; Fine-tuned version of ProsusAI/finbert. <br>&bull; Fine-tuned on data useed in FOMC minutes from Jan 2006 to Feb 2023.                                    |
| [finbert-sentiment-analysis](https://huggingface.co/AventIQ-AI/finbert-sentiment-analysis) | AventIQ-AI/finbert-sentiment-analysis | aventiq_ai  | &bull; FinBERT model fine-tuned for sentiment analysis using English/Quotes dataset.                                                                             |

## Observations

- In general, FinBERT is able to rate the sentiment correctly to a certain extent.
- Different variants perform differently based on the fine-tuned dataset - 262 news (out of 3791) are rated as both negative and positive concurrently based on 28 March analysis.
- 'FinBERT-FOMC' appears to have more consistent rating among the 4 (Needs further investigations).


In [2]:
df = pd.read_csv("./data/sentiment.csv")
df

pub_date ticker        publisher          period  \
0     2025-03-28 18:53:00   AAPL    Yahoo Finance     2 hours ago   
1     2025-03-28 20:08:00   AAPL  Simply Wall St.  45 minutes ago   
2     2025-03-28 19:53:00   AAPL     MT Newswires      1 hour ago   
3     2025-03-28 19:53:00   AAPL      Motley Fool      1 hour ago   
4     2025-03-28 18:53:00   AAPL         LA Times     2 hours ago   
...                   ...    ...              ...             ...   
3788  2025-03-04 21:02:00     BA    The Telegraph     24 days ago   
3789  2025-03-03 21:02:00     BA  PA Media: Money     25 days ago   
3790  2025-03-03 21:02:00     BA          Reuters     25 days ago   
3791  2025-03-01 21:02:00     BA      Motley Fool     27 days ago   
3792  2025-03-01 21:02:00     BA          Reuters     27 days ago   

                                                  title  \
0     Trump is set on tariffing chips. It’s not so s...   
1     Cloud AI Today - Boosting Performance Akamai a...   
2     Market Chatter: Apple, Meta Platforms Could be...   
3     Prediction: This Artificial Intelligence (AI) ...   
4     Expansive offices for Apple are rising in Culv...   
...                                                 ...   
3788  British Airways opens new Gatwick hangar to cu...   
3789  Aerospace giant Senior shrugs off Boeing strik...   
3790         Norwegian to buy 10 leased Boeing aircraft   
3791  Boeing's Space Business Could Be Stoppable Aft...   
3792  Spirit AeroSystems reports $577 million quarte...   

                                                content  prosusai  \
0     President Trump's plan to put 25% tariffs on c...         3   
1     Fermyon Technologies and Akamai have announced...         5   
2     Apple (AAPL) and Meta Platforms (META) could b...         2   
3     The trillion-dollar club could be joined by an...         3   
4     Construction crews have erected cranes and are...         4   
...                                                 ...       ...   
3788  British Airways is to open a new maintenance h...         4   
3789  The UK firm, which is based in Rickmansworth, ...         1   
3790  COPENHAGEN (Reuters) -Budget carrier Norwegian...         5   
3791  I was wrong about Boeing in December. Imagine ...         3   
3792  Boeing and European rival Airbus -- already ag...         2   

      prosusai_title  prosusai_content  yiyanghkust  yiyanghkust_title  \
0                  3                 3            3                  3   
1                  5                 5            5                  5   
2                  1                 3            3                  3   
3                  3                 3            3                  3   
4                  5                 4            5                  3   
...              ...               ...          ...                ...   
3788               4                 3            3                  3   
3789               2                 1            2                  1   
3790               3                 5            5                  3   
3791               3                 3            1                  1   
3792               3                 4            1                  1   

      yiyanghkust_content  ziweichen  ziweichen_title  ziweichen_content  \
0                       3          3                3                  2   
1                       5          5                5                  5   
2                       3          2                2                  2   
3                       3          3                3                  3   
4                       5          5                3                  5   
...                   ...        ...              ...                ...   
3788                    3          3                3                  3   
3789                    3          1                3                  3   
3790                    5          3                3                  3   
3

In [3]:
print(df["ticker"].nunique())
df["ticker"].unique()

19


array(['AAPL', 'NVDA', 'MSFT', 'AMZN', 'GOOGL', 'META', 'TSLA', 'JPM',
       'JNJ', 'V', 'XOM', 'UNH', 'WMT', 'PG', 'HD', 'NFLX', 'CRM', 'BAC',
       'BA'], dtype=object)

In [ ]:
utils.display_divergent_rating(df)

divergent news : 262


# [Trade Results](#table-of-contents)


In [105]:
results = utils.load_csv("./data/results/2025-03-28/trade_results.csv")
results

ticker coint_ticker action  entry_date entry_price   exit_date  \
0        HD          MSI    buy  2025-01-28      464.02  2025-03-06   
1        HD          MSI    buy  2025-02-26      425.15  2025-03-06   
2        HD          MSI    buy  2025-02-27      428.42  2025-03-06   
3        HD          MSI    buy  2025-02-28      439.06  2025-03-06   
4        HD          MSI    buy  2025-03-03      426.23  2025-03-06   
...     ...          ...    ...         ...         ...         ...   
1166     BA        CMCSA    buy  2025-03-12       35.19  2025-03-13   
1167     BA        CMCSA    buy  2025-03-14       35.43  2025-03-19   
1168     BA        CMCSA    buy  2025-03-18       36.01  2025-03-19   
1169     BA        CMCSA    buy  2025-03-20        36.5  2025-03-25   
1170     BA        CMCSA    buy  2025-03-26       37.37  2025-03-27   

     exit_price profit_loss  days_held  win  
0         422.2      -41.82         37    0  
1         422.2       -2.95          8    0  
2         422.2       -6.22          7    0  
3         422.2      -16.86          6    0  
4         422.2       -4.03          3    0  
...         ...         ...        ...  ...  
1166       34.7       -0.49          1    0  
1167      36.09        0.66          5    1  
1168      36.09        0.08          1    1  
1169      36.94        0.44          5    1  
1170      37.37         0.0          1    0  

[1171 rows x 10 columns]

# [Overall Summary](#table-of-contents)


In [16]:
overall = utils.load_csv("./data/results/2025-03-28/overall_summary.csv", index_col=0)
overall

Overall Statistics
stock_ticker_without_trades                                               AAPL
stock_ticker_with_trades     HD, CRM, TSLA, AMZN, V, NVDA, BA, JNJ, PG, JPM...
total_num_trades                                                          1171
total_num_wins                                                             597
total_num_loss                                                             574
total_profit                                                           -293.94
max_loss                                                                -84.56
max_profit                                                               32.61
mean_profit                                               -0.25101622544833474
median_profit                                                             0.02
total_investment                                                     194224.93
first_trade                                                         2025-01-28
last_trade                                                          2025-03-26
trading_period                                                              58
percent_return                                                       -0.001513
annualized_return                                                    -0.009483

# [Breakdown Summary](#table-of-contents)


In [101]:
breakdown = utils.load_csv(
    "./data/results/2025-03-28/breakdown_summary.csv", header=[0, 1], index_col=[0, 1]
)
breakdown

entry_date               exit_date days_held            \
                            min         max         max       min max  mean   
ticker coint_ticker                                                           
AMZN   BAC           2025-03-20  2025-03-26  2025-03-27         1   7  4.25   
       CI            2025-03-20  2025-03-26  2025-03-27         1   7  4.25   
       HII           2025-03-20  2025-03-26  2025-03-27         1   7  4.25   
       HSY           2025-03-20  2025-03-26  2025-03-27         1   7  4.25   
       INVH          2025-03-20  2025-03-26  2025-03-27         1   7  4.25   
...                         ...         ...         ...       ...  ..   ...   
XOM    GOOG          2025-02-26  2025-03-26  2025-03-27         1  10   4.7   
       GOOGL         2025-02-26  2025-03-26  2025-03-27         1  10   4.7   
       SJM           2025-02-26  2025-03-26  2025-03-27         1  10   4.7   
       WSM           2025-02-26  2025-03-26  2025-03-27         1  10   4.7   
       WY            2025-02-26  2025-03-26  2025-03-27         1  10   4.7   

                           entry_price profit_loss                         \
                    median         sum         min    max    mean  median   
ticker coint_ticker                                                         
AMZN   BAC             4.5      170.84       -0.51   0.09   -0.15   -0.09   
       CI              4.5     1279.67        3.35   8.09  4.8225   3.925   
       HII             4.5      812.34       -2.74   1.68   0.065    0.66   
       HSY             4.5      668.18         2.1   6.15   3.575   3.025   
       INVH            4.5      136.52       -0.21   0.29    0.03    0.02   
...                    ...         ...         ...    ...     ...     ...   
XOM    GOOG            3.5     1675.28       -4.48   3.58  -0.896  -0.845   
       GOOGL           3.5     1654.83       -4.22   3.32  -0.953    -0.9   
       SJM             3.5     1114.87       -4.56   2.25  -0.901   -1.25   
       WSM             3.5     1752.93      -11.35  -0.94  -6.299   -6.47   
       WY              3.5      296.13       -0.93   1.28  -0.126   -0.19   

                              win     trading_period percent_ret annual_ret  
                        sum count sum                                        
ticker coint_ticker                                                          
AMZN   BAC             -0.6     4   2              7   -0.003512   0.832393  
       CI             19.29     4   4              7    0.015074   2.181792  
       HII             0.26     4   2              7     0.00032   1.016826  
       HSY             14.3     4   4              7    0.021401   3.016682  
       INVH            0.12     4   2              7    0.000879   1.046879  
...                     ...   ...  ..            ...         ...        ...  
XOM    GOOG           -8.96    10   4             29   -0.005348   0.934731  
       GOOGL          -9.53    10   3             29   -0.005759   0.929887  
       SJM            -9.01    10   3             29   -0.008082   0.902912  
       WSM           -62.99    10   0             29   -0.035934   0.630905  
       WY             -1.26    10   4             29   -0.004255   0.947748  

[177 rows x 18 columns]

In [3]:
tsla_cag = pd.read_csv(
    "./data/results/2025-03-28/ziweichen_5/price_actions/TSLA_CAG.csv"
)
tsla_cag

date ticker   day_name  is_holiday  av_rating  median_rating_excl  \
0  2025-03-25   TSLA    Tuesday       False   2.764706                 1.5   
1  2025-03-26   TSLA  Wednesday       False   3.163636                 4.5   
2  2025-03-27   TSLA   Thursday       False   2.800000                 2.0   
3  2025-03-28   TSLA     Friday       False   2.530612                 1.0   

   TSLA_close  CAG_close action  
0      288.14      25.35   sell  
1      272.06      25.92    buy  
2      273.13      26.46   sell  
3         NaN        NaN   sell

# [Future Works](#table-of-contents)

- Determine if news title suffice in sentiment analysis.
- Determine pvalue threshold for cointegration studies (Is lowest pvalue = highest profit)
- Abberation studies using profit/loss generated based on sentiment rating:

| S/N | Component            | Description                                                                                                               |
| --- | -------------------- | ------------------------------------------------------------------------------------------------------------------------- |
| 1.  | FinBert model        | &bull; Prosus AI, finbert-tone, FinBERT-FOMC, finbert-sentiment-analysis. <br>&bull; Currently tested on 'FinBERT-FOMC'.  |
| 2.  | News info            | &bull; news title, news content, both news title and content. <br>&bull; Currently tested on both news title and content. |
| 3.  | cointegration        | &bull; 1, 3 and 5 years. <br>&bull; Currently tested on 5 years.                                                          |
| 4.  | correlation analysis | &bull; Pearson (Linear correlation), Spearman & Kendall (monotonic correlation). <br>&bull; Currently not tested.         |
